### DISCUSSION TITLE
rendering CSS

### DISCUSSION THREAD

**Discussion by gwydion67** [vriteio/vrite#23] [Ideas]
I have been using joplin till now, and the biggest reason why I can't change is it's ability to render CSS. I really like the work and design of the app you build so i was wishing if you could try and add the feature to render CSS. sweat_smile

**Comment by areknawo** [comment]
What exactly do you mean by "render CSS"?

If you're referring to support for CSS in the editor then the integrated code editor supports CSS syntax highlighting and even autocompletions.

On the other hand, if you want to customize how the content is delivered then you need to remember Vrite is a **headless** CMS. This means it doesn't limit you to a ready frontend, and helps you build your own website, providing access to the content from the Vrite API.

That said, possibly you are referring to something entirely different, in which case I'd need a bit more details. I'm not familiar with Joplin.

**Comment by gwydion67** [comment]
![image](https://github.com/vriteio/vrite/assets/93286089/8405760d-b3a4-4f05-96b9-ab07ca774867)
this is what i meant by rendering CSS. this CSS is applied to the files at export as well

**Comment by areknawo** [comment]
Oh. That's quite an interesting feature.

Unfortunately this doesn't quite fit with Vrite's current design and underlying content format (JSON). The goal is to have a clean format that's easy to process and deliver to various endpoints. CSS doesn't really fit into the mix.

As an alternative, a **custom block and inline content** is planned. This would allow you to add new options to the editor via extensions, which have a customizable in-editor look and output standardized JSON, compatible with the built-in elements.

Coming from Joplin, I assume you're using this for note taking. In this case the approach might even be preferable. You could create (or find) an extension that provides inline e.g. `Important` formatting which styles the text red in the editor and outputs e.g.:

```json
{
    "type": "text",
    "text": "is hardly used nowadays",
    "marks": [{ "type": "important", "attrs": {} }]
}
```

Possibly even a `Styled` formatting could be possible, which lets you set entirely custom style and output it as an attribute, e.g.

```json
{
    "type": "text",
    "text": "is hardly used nowadays",
    "marks": [{ 
       "type": "styled",
       "attrs": {
         "style": "color:red;"
       }
    }]
}
```

Can't really say if this will be feasable at this moment.

That said, for note-taking you'll likely care primarly about the visuals in the editor, which custom content extensions will allow you to customize (with CSS). This won't be as flexible as custom CSS support right in the editor, but should work just as well (if not even better) for majority of cases.

**Comment by areknawo** [comment]
Will close this discussion for now. As I've said, I plan to implement custom block elements and public Vrite Extensions for the stable release. This might just support your use case in a different way. Built-in support for rendering CSS is not planned (at least for the foreseeable future).

Feel free to reopen if you have further feedback.
Thanks!
